enhancements/authentication/pod-security-admission-enforcement.md

```diff
@@ -0,0 +1,428 @@
+
+`status.violatingNamespaces` lists the Namespaces that would fail if `status.enforcementMode` were `LabelEnforcement` or `FullEnforcement`.
+The reason field helps identify whether the PSA label syncer or the PodSecurity config is the root cause.
+Administrators must query the kube-apiserver (or use the [cluster debugging tool](https://github.com/openshift/cluster-debug-tools)) to pinpoint specific workloads.
+
+### Implementation Details
+
+- The `PodSecurityReadinessController` in the `cluster-kube-apiserver-operator` will manage the new API.
+- If the `FeatureGate` is removed from the current `FeatureSet`, the cluster must revert to its previous state.
+- The [`Config Observer Controller`](https://github.com/openshift/cluster-kube-apiserver-operator/blob/218530fdea4e89b93bc6e136d8b5d8c3beacdd51/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L131) must be updated to watch for the new API alongside the `FeatureGate`.
+
+#### PodSecurityReadinessController
+
+The `PodSecurityReadinessController` will manage the `PSAEnforcementConfig` API.
+It already collects most of the necessary data to determine whether a Namespace would fail enforcement or not to create a [`ClusterFleetEvaluation`](https://github.com/openshift/enhancements/blob/master/dev-guide/cluster-fleet-evaluation.md).
+With the `security.openshift.io/MinimallySufficientPodSecurityStandard`, it will be able to evaluate all Namespaces for failing workloads, if any enforcement would happen.
+With the `security.openshift.io/ValidatedSCCSubjectType`, it can categorize violations more accurately and create a more accurate `ClusterFleetEvaluation`.
+
+#### PodSecurity Configuration
+
+A Config Observer in the `cluster-kube-apiserver-operator` manages the Global Config for the kube-apiserver, adjusting behavior based on the feature flag.
+It must watch both the `status.enforcementMode` and the `FeatureGate` to make decisions.
+
+#### PSA Label Syncer
+
+The PSA label syncer will watch the `status.enforcementMode` and the `OpenShiftPodSecurityAdmission` feature gate.
+If `status.enforcementMode` is `LabelEnforcement` or `FullEnforcement` and `OpenShiftPodSecurityAdmission` is enabled, the syncer will set the `pod-security.kubernetes.io/enforce` label.
+Otherwise, it will refrain from setting that label and remove any enforce labels it owns if existent.
+
+Because the ability to set `pod-security.kubernetes.io/enforce` is introduced in release `n`, the ability to remove that label must exist in release `n-1`.
+Otherwise, the cluster will be unable to revert to its previous state.
+
+## Open Questions
+
+### Fresh Installs
+
+Needs to be evaluated. The System Administrator needs to pre-configure the new API’s `spec.targetMode`, choosing whether the cluster will be `privileged`, `restricted`, or `conditional` during a fresh install.
+
+### Impact on HyperShift
+
+Needs to be evaluated.
+
+### Baseline Clusters
```

Comment:
Moved here:

> Is there any particular reason there isn't an option for the Baseline enforcement level? Is this enforcement level one that users may want to configure?

*https://github.com/openshift/enhancements/pull/1747/files#r1941774952*

Reply:
Yeah, we could do that. It technically doubles the effort, but `O(n) == O(2n)`, right? :D 

We would check the cluster for namespaces that have no enforce label and try restricted upon them. If it fails we would try baseline. If all of the reminding namespaces support at least baseline, we could move to baseline.

I am hesitant, as across the whole reviewing section to add different states to reason about, but this is a minimal overhead for a significant gain.

Reply:
Could we just allow `Baseline` as a configuration option? I don't think we should automagically decide to set the global config to `baseline` if the desired state of `Restricted` is not achievable but `baseline` is.

Reply:
Decided against it in the Arch Call.